flytekit/core/type_engine.py

```diff
@@ -463,6 +463,10 @@ def _fix_val_int(self, t: typing.Type, val: typing.Any) -> typing.Any:
             # Handle nested Dict. e.g. {1: {2: 3}, 4: {5: 6}})
             return {self._fix_val_int(ktype, k): self._fix_val_int(vtype, v) for k, v in val.items()}
 
+        if get_origin(t) is typing.Union and type(None) in get_args(t):
+            # Handle optional type. e.g. Optional[int], Optional[dataclass]
+            return self._fix_val_int(get_args(t)[0], val)
```

Comment:
Does this work if the user doesn't do Optional, but instead does `typing.Union[int, None]`.

Also unrelated, can we add a unit test to document what happens in the case of `typing.Union[int, float]` (where the value is like 5 or something, something that's an int?

Reply:
Marshmellow doesn't support union type. https://github.com/marshmallow-code/marshmallow/issues/1191
Therefore, we can't use `typing.Union[int, float]` in dataclass. btw, `Union[None, int]` is also an optional type, but Marshmallow does not support it.

will also update the doc of the dataclass example in flytesnack.